testing/test_stepper_control.py

```diff
@@ -0,0 +1,185 @@
+            self._mock_motor.step.side_effect = lambda: self.assertFalse(self._mock_motor.clockwise,
+                                                                         msg="Shouldn't be stepping clockwise!")
+
+        self._mock_motor.step.side_effect = press_switch_on_first_step_then_assert_anticlockwise
+
+        # Test that we do not step further
+        try:
+            for i in range(10):
+                self._axis.step()
+        except UnexpectedLimitSwitchError:
+            # We expect this error!
+            pass
+
+    def test_current_location_shows_2mm_backoff_after_a_limit_switch_is_pressed(self):
+        """Only 2mm since then if we go the wrong way, we have not gone through the whole travel of the switch."""
+
+        def press_limit_switch():
+            self._mock_limit_switches[0].is_pressed = True
+
+        for i in range(10):
+            self._axis.step()
+
+        collision_location = self._axis.current_location
+        if self._axis.forwards:
+            expected_backoff_location = collision_location - 2
+        else:
+            expected_backoff_location = collision_location + 2
+
+        self._mock_motor.step.side_effect = press_limit_switch
+
+        try:
+            self._axis.step()
+        except UnexpectedLimitSwitchError:
+            pass
+
+        self.assertAlmostEqual(self._axis.current_location, expected_backoff_location,
+                               delta=self._axis.millimetres_per_step)
+
+    def test_raises_even_when_overridden_if_switch_is_still_pressed_after_backoff(self):
+        def press_limit_switch():
+            self._mock_limit_switches[0].is_pressed = True
+
+        self._mock_motor.step.side_effect = press_limit_switch
+        self._axis.override_limit_switches = True
+        with self.assertRaises(UnexpectedLimitSwitchError):
+            self._axis.step()
+
+    def test_stepping_raises_when_a_limit_switch_is_pressed(self):
+        for i in 0, 1:
+            self._reset_switches()
+
+            def press_limit_switch():
+                self._mock_limit_switches[i].is_pressed = True
+
+            self._mock_motor.step.side_effect = press_limit_switch
+
+            with self.assertRaises(UnexpectedLimitSwitchError):
+                self._axis.step()
+
+    def _reset_switches(self):
+        for switch in self._mock_limit_switches:
+            switch.is_pressed = False
+
+
+class AxisHomingTest(BaseTestCases.Axis):
+    """Tests the behaviour of the Axis.home() method."""
+
+    def setUp(self):
+        super().setUp()
+        self.true_motor_location_in_steps = 20  # Some non-zero start location
+        self._mock_motor.step.side_effect = self._default_motor_side_effect
+
+    def _default_motor_side_effect(self):
+        # Advance the 'real' location (in steps)
+        if self._mock_motor.clockwise:
+            self.true_motor_location_in_steps += 1
+        else:
+            self.true_motor_location_in_steps -= 1
+
```

Comment:
Tests are generally looking good. More comments would be good as I got confused in places especially around some of the helper functions. But as these are tests and I think I worked most of it out, just bear in mind for future.